src/crystal/hasher.cr

```diff
@@ -45,12 +94,53 @@ struct Crystal::Hasher
   end
 
   def reference(value)
-    @result = @result * 31 + value.object_id.to_u64
-    self
+    permute(value.object_id.to_u64)
   end
 
   def string(value)
-    value.to_slice.hash(self)
+    bytes(value.to_slice)
+  end
+
+  private def read_u24(u, r)
+    u[0].to_u64 | (u[r/2].to_u64 << 8) | (u[r - 1].to_u64 << 16)
+  end
+
+  private def read_u32(u)
+    # force correct unaligned read
+    t4 = uninitialized UInt32
+    pointerof(t4).as(UInt8*).copy_from(u, 4)
+    t4.to_u64
+  end
+
+  private def read_u64(u)
+    # force correct unaligned read
+    t8 = uninitialized UInt64
+    pointerof(t8).as(UInt8*).copy_from(u, 8)
+    t8
+  end
+
+  def bytes(value)
+    bsz = value.size
+    u = value.to_unsafe
+    if bsz == 0
```

Comment:
One small suggestion:

```cr
case bsz
...
end
```

Reply:
Sorry, but with `case` it is consistently a bit slower.
Main cause is `Slice@size` is signed. With `if` optimizer thinks, that all negative values are handled in `if bsz <= 3`, so that it correctly predicts that `bsz >= 8` always before loop. With `case bsz; when 0; when 1,2,3; when 4,5,6,7; else; end` optimizer can not make such assumption.
But even if I cast bsz to UInt32, `case` is still a bit slower.